bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/operations/OperationHistoryActionHandler.java

```diff
@@ -298,7 +303,32 @@ public final void run() {
 			if (getOperation() instanceof IAdvancedUndoableOperation2) {
 				runInBackground = ((IAdvancedUndoableOperation2) getOperation()).runInBackground();
 			}
-			progressDialog.run(runInBackground, true, runnable);
+			if (runInBackground) {
+				progressDialog.run(runInBackground, true, runnable);
+			} else {
+				final IJobManager manager = Job.getJobManager();
+				final ISchedulingRule rule = (getOperation() instanceof ISchedulableOperation)
+						? ((ISchedulableOperation) getOperation()).getSchedulingRule()
+						: null;
+				if (rule == null) {
+					progressDialog.run(runInBackground, true, runnable);
+				} else {
+					// prevent UI freeze during AutoBuild as in
+					// org.eclipse.jdt.internal.ui.refactoring.RefactoringExecutionHelper#perform
+					try {
+						try {
+							// interrupt autobuild or show Wait/Cancel Dialog:
+							Runnable r = () -> manager.beginRule(rule, null);
+							BusyIndicator.showWhile(parent.getDisplay(), r);
```

Comment:
@laeubi : you've mentioned on https://github.com/eclipse-platform/eclipse.platform.ui/pull/16#issuecomment-1102520698
>  Anyways I think the problem is not necessarily a lock here but the BusyIndicator and progressDialog.run both need to be handled with care and might not behave as expsected as when not running in the background require that the caller make sure to drive the eventqueue!

Can you please review **this** change?

Reply:
@iloveeclipse sadly it is not documented well but the [snippet for the BusyIndicator](https://github.com/eclipse-platform/eclipse.platform.swt/blob/master/examples/org.eclipse.swt.snippets/src/org/eclipse/swt/snippets/Snippet130.java) shows that:

1. You should spin a new thread to actually do the really long running work
2. busy-loop to check if that spinned thread is done and additionally drive the event queue in the runnable that actually is passed to the BusyIndicator

otherwise you will see the busy-cursor but no event will be processed what could then be reported as a "this application seems to be blocked" and of course any Display.async/sync calls to block / deadlock.

I don't know if the manager.begin rule do all of that internally.

Because of this complex and setup I once proposed a change to SWT for that I think but it was not included so I'm using a helper method for this in my own code that uses CompleteableFutures to help with that.
If we thing its now a better time for such a change I can try to continue my past work on that but thi will require some additions in SWT first.

Reply:
`JobManager.beginRule()` has no understanding of UI at all, and no hook to make UI react to rule changes; that's one reason why such UI freeze because of scheduling rules can happen. As I suggest in https://github.com/eclipse-platform/eclipse.platform.ui/issues/8 , some way to hook more UI friendliness in JobManager would allow to automatically show a progressDialog or a busyCursor in case a rule is requested in UI Thread and isn't available immediately.

Reply:
It might or might not be a problem if the begin rule is fast, but then a BusyIndicator seems not valid to use, also the comment seems confusing, as a busy indicator can't be used for cancel ...

Reply:
I did not follow this conversation: do you request any change in this piece of code? Its copied pattern from RefactoringExecutionHelper#perform. Its tested, It's working. - can we resolve this?

Reply:
Just because it is used else where do not mean it correct, please check the `BusyIndicator` implementation, there are two cases:

1. You are in the UI thread then the UI thread will **block the UI** until the runnable is done (even though showing a busy cursor) unless you drive the event queue.
2. You are not in the UI thread and then the runnable is just running (still blocking) without any busy cursor.

Reply:
and what do you suggest from that? we are in UI and it is blocking. but its not freezing anymore and its showing a cancel dialog.

Reply:
BusyIndicator has nothing to do with cancel or not freezing, BusyIndicator **will freeze** the UI if not used appropriate (see the linked code snippet on how it is supposed to be used).

Reply:
don't trust a snippet but actually try it!

Reply:
Maybe you want to try it out yourself :-)

Just set a breakpoint at `manager.beginRule(rule, null)` you will see that there is a busy cursor and your Application will be marked as "Non responsive", and no cancel option.

But well no problem if you think its fine, it was just requested by @iloveeclipse to give a hint about a possible dead. lock.

Reply:
please read the "how to reproduce" in https://bugs.eclipse.org/bugs/show_bug.cgi?id=578871#c0
it's not about putting a breakpoint in the UI thread but about putting a breakpoint in the autobuild thread. of cause the UI thread needs to run in order to show a dialog.

Reply:
Well there are two cases:

1. `manager.beginRule(rule, null)` then there is no point in using BusyIndicator
2. `manager.beginRule(rule, null)` could sometimes be slow (e.g waiting for a lock from another thread) then this will freeze the UI, this has nothing to do with whatever has to be fixed/reproduced with this.

Reply:
I expressed my concerns and give you hints where blocking occurs, if you think it is invalid I'm fine with that.